docs/contributing/muzzle.md

```diff
@@ -0,0 +1,32 @@
+# Muzzle
+
+Muzzle is a feature of the Java agent that ensures API compatibility
+between libraries/symbols on the application classpath and APIs of instrumented
+3rd party libraries used by the Agent. In other words the Muzzle ensures
+that the API symbols used by the Agent are compatible with API symbols
+on the application classpath. The Muzzle will prevent loading an instrumentation
+if the APIs do not match.
+
+## How does it work
+
+At build time the Muzzle gradle plugin generates a matcher for 3rd party APIs used by the agent.
+The matcher is stored in the instrumentation class in method `ReferenceMatcher getInstrumentationMuzzle()`.
+
+At runtime the Muzzle checks API compatibility between symbols used by the Agent
+and symbols in the application class loader. Before inspecting the class loader
+the Muzzle plugin checks symbols defined in `SomeInstrumentation.typeMatcher()`
+as a performance optimization. If the symbols do not match the instrumentation
+is not loaded.
+
+## Muzzle gradle plugin
+
+The `printReferences` task prints 3rd party API references in a given module.
+
+```bash
+./gradlew :instrumentation:google-http-client-1.19:printReferences
+```
+
+## Muzzle location
+
+* `buildSrc` - Muzzle Gradle plugin
```

Comment:
This part of muzzle isn't really documented above.  This plugin is responsible for downloading the range of libraries from Maven Central and verifying expected compatibility.  If a new incompatible version is published, this is what will break the build.

Reply:
Is there also a task to run it explicitly?

Reply:
Yeah, this is what actually runs when you call the `muzzle` task.
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/.circleci/config.yml#L132

That can be called at the top level, or for an individual project.
As an optimization, these tasks are only added to gradle if `muzzle` is explicitly listed as a task, otherwise it slows down the config stage substantially.